specification/dartLangSpec.tex

```diff
@@ -22093,108 +24137,335 @@ \subsubsection{Actual Types}
+\LMHash{}%
+This specification does not yet specify type inference in Dart.\!\footnote{%
+A preliminary specification is available at
+\url{https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md}.%
+}
+A future version of this specification will include that topic.
 
 \LMHash{}%
-% does this diverge in some cases?
-Given two interfaces $I$ and $J$,
-let $S_I$ be the set of superinterfaces of $I$,
-let $S_J$ be the set of superinterfaces of $J$
-and let $S = (\{I\} \cup S_I) \cap (\{J\} \cup S_J)$.
-Furthermore,
-we define $S_n = \{T | T \in S \wedge depth(T) = n\}$ for any finite $n$
-where $depth(T)$ is the number of steps in the longest inheritance path
-from $T$ to \code{Object}.
-%TODO(lrn): Specify: "inheritance path" is a path in the superinterface graph.
-Let $q$ be the largest number such that $S_q$ has cardinality one,
-which must exist because $S_0$ is $\{\code{Object}\}$.
-The least upper bound of $I$ and $J$ is the sole element of $S_q$.
+In other parts of this specification,
+type inference is generally assumed to have taken place already.
+However, commentary is used in several locations to give some informal
+information about the effects of type inference.
+
+
+\section{Flow Analysis}
+\LMLabel{flowAnalysis}
 
+%% TODO(eernst): Come flow analysis, rewrite this section entirely.
 \LMHash{}%
-The least upper bound of \DYNAMIC{} and any type $T$ is \DYNAMIC.
-The least upper bound of \VOID{} and any type $T \ne \DYNAMIC{}$ is \VOID.
-The least upper bound of $\bot$ and any type $T$ is $T$.
-Let $U$ be a type variable with upper bound $B$.
-The least upper bound of $U$ and a type $T \ne \bot$ is
-the least upper bound of $B$ and $T$.
+This specification does not fully specify the flow analysis in Dart,
+but a future version of it will do so.\!\footnote{%
+A preliminary specification is available at
+\url{https://github.com/dart-lang/language/blob/master/resources/type-system/flow-analysis.md}.%
+}
+In this section we specify a few concepts and special cases
+in the flow analysis
+that do not depend on the detailed control flow,
+so as to enable the rest of the specification to refer to information
+obtained during this analysis.
 
 \LMHash{}%
-The least upper bound operation is commutative and idempotent,
-but it is not associative.
+The flow analysis provides information about reachability of code,
+about the definite assignment status of each local variable,
+and about type test related events in the control flow
+that determine type promotion of local variables.
+
 
-% Function types
+\subsection{Definite Assignment}
+\LMLabel{definiteAssignment}
+
+\LMHash{}%
+At each location where a local variable can be referred
+(\commentary{%
+e.g., as an expression, or as the left hand side of an assignment%
+}),
+the variable has a status as being
+\IndexCustom{definitely assigned}{local variable!definitely assigned} or
+\IndexCustom{definitely unassigned}{local variable!definitely unassigned}.
```

Comment:
This clause sounds as if it's strictly one of these two possibilities, but as stated below, a variable can be neither definitely assigned nor definitely unassigned.